src/MagnumPlugins/CgltfImporter/CgltfImporter.cpp

```diff
@@ -0,0 +1,2562 @@
+    if(buffer.data)
+        return true;
+
+    Containers::ArrayView<const char> view;
+    if(!buffer.uri) {
+        /* URI may only be empty for buffers referencing the glb binary blob */
+        if(id != 0 || !data->bin) {
+            Error{} << "Trade::CgltfImporter::" << Debug::nospace << function << Debug::nospace << "():" <<
+                "buffer" << id << "has no URI";
+            return false;
+        }
+        view = Containers::arrayView(static_cast<const char*>(data->bin), data->bin_size);
+    } else {
+        const auto loaded = loadUri(buffer.uri, bufferData[id], function);
+        if(!loaded)
+            return false;
+        view = *loaded;
+    }
+
+    /* The spec mentions that non-GLB buffer length can be greater than
+       byteLength. GLB buffer chunks may also be up to 3 bytes larger than
+       byteLength because of padding. So we can't check for equality. */
+    if(view.size() < buffer.size) {
+        Error{} << "Trade::CgltfImporter::" << Debug::nospace << function << Debug::nospace << "():" <<
+            "buffer" << id << "is too short, expected" << buffer.size <<
+            "bytes, but got" << view.size();
+        return false;
+    }
+
+    buffer.data = const_cast<char*>(view.data()); /* sigh */
+    /* Tell cgltf not to free buffer.data in cgltf_free */
+    buffer.data_free_method = cgltf_data_free_method_none;
+
+    return true;
+}
+
+Containers::Optional<Containers::StridedArrayView2D<const char>> CgltfImporter::Document::accessorView(const cgltf_accessor* accessor, const char* const function) {
+    /* All this assumes the accessor was checked using checkAccessor() */
+    const cgltf_buffer_view* bufferView = accessor->buffer_view;
+    const cgltf_buffer* buffer = bufferView->buffer;
+    const UnsignedInt bufferId = buffer - data->buffers;
+    if(!loadBuffer(bufferId, function))
+        return Containers::NullOpt;
+
+    return Containers::StridedArrayView2D<const char>{Containers::arrayView(buffer->data, buffer->size),
+        reinterpret_cast<const char*>(buffer->data) + bufferView->offset + accessor->offset,
+        {accessor->count, elementSize(accessor)},
+        {std::ptrdiff_t(accessor->stride), 1}};
+}
+
+Containers::StringView CgltfImporter::Document::decodeString(Containers::StringView str) {
+    if(str.isEmpty())
+        return str;
+
+    /* String has been decoded before */
+    const auto found = decodedStrings.find(str.data());
+    if(found != decodedStrings.end())
+        return found->second;
+
+    /* The input string can be UTF-8 encoded but we can use a byte search here
+       since all multi-byte UTF-8 characters have the high bit set and '\\'
+       doesn't, so this will only match single-byte ASCII characters. */
+    Containers::StringView escape = str.find('\\');
+    /* No escaped sequence found. If the view is null-terminated (all strings
+       in cgltf_data should be), this stores a non-owning String. */
```

Comment:
It does allocate a null-terminated copy, yes. And there's no convenient way to get the original pointer to json memory that I can see. There are two alternatives to the current implementation:

1. Decode all strings we could possibly need later. Not a fan of imposing that overhead at import time.
2. Same as now, but decode strings in-place on cgltf-allocated strings (or even `delete[]` and replace). Saves some memory but we'd still have to track what's already decoded, maybe with an `unordered_set`.

I went with the `unordered_map` because it felt cleaner to abstract (not really true), but I'm starting to like option 2 a bit more, if just for the memory savings. Opinions?

Reply:
I was just wondering why the design decision in cgltf / jsmn went like that, why it doesn't go all the way and unescape as well, or why on the other hand it doesn't just keep pointing to the original data if it's unescaped anyway. Doing a copy just to make them null-terminated feels silly tbh :)

Your implementation on the plugin side is okay, I wouldn't change that, for now at least. If it ever becomes a bottleneck, I'd instead look into fixing this directly in cgltf or jsmn (not sure which of the two is responsible).

Reply:
jsmn doesn't perform any allocation, so this is cgltf's doing. Makes it a bit easier to fix things upstream if we ever need that, but I suspect removing null-terminated strings wouldn't be easy to get merged in any case.